### DISCUSSION TITLE
Cannot build simplest alire program in MacOS

### DISCUSSION THREAD

**Discussion by Darce42** [alire-project/alire#1471] [Support] [answered]
Greetings All
I'm hoping that some kind person can tell me why I'm unable to build the simplest Alire (alr) project.
System: imac 27 inch, Intel core i9, 40GB RAM, OS is Sonara 14.0.
I believe everything is up to date, including alire. 

I've created an example using alr init (the simplest possible with just a null; in the program), named mycrate in the Temp directory.
This fails to build as do other earlier programs I have, that is, programs that used to build correctly.
Something has changed but I'm not sure what.

Here is the result of alr build on that trivial program.

thanks in advance

ⓘ Building mycrate/mycrate.gpr...
Setup
   [mkdir]        object directory for project Mycrate
   [mkdir]        exec directory for project Mycrate
Compile
   [Ada]          mycrate.adb
Bind
   [gprbind]      mycrate.bexch
   [Ada]          mycrate.ali
Link
   [link]         mycrate.adb
-macosx_version_min has been renamed to -macos_version_min
0  0x10d884f43  __assert_rtn + 64
1  0x10d786f43  ld::AtomPlacement::findAtom(unsigned char, unsigned long long, ld::AtomPlacement::AtomLoc const*&, long long&) const + 1411
2  0x10d7a3431  ld::InputFiles::SliceParser::parseObjectFile(mach_o::Header const*) const + 19745
3  0x10d7b3e44  ld::InputFiles::parseAllFiles(void (ld::AtomFile const*) block_pointer)::$_7::operator()(unsigned long, ld::FileInfo const&) const + 1380
4  0x7ff80daef5cd  _dispatch_client_callout2 + 8
5  0x7ff80daffe3e  _dispatch_apply_invoke + 214
6  0x7ff80daef59a  _dispatch_client_callout + 8
7  0x7ff80dafe99d  _dispatch_root_queue_drain + 879
8  0x7ff80dafef22  _dispatch_worker_thread2 + 152
9  0x7ff80dc93c06  _pthread_wqthread + 262
ld: Assertion failed: (resultIndex < sectData.atoms.size()), function findAtom, file Relocations.cpp, line 1336.
collect2: error: ld returned 1 exit status
gprbuild: link of mycrate.adb failed
gprbuild: failed command was: /users/imac27/.config/alire/cache/dependencies/gnat_native_12.2.1_77267eb1/bin/gcc mycrate.o b__mycrate.o -L/Users/imac27/Documents/Temp/mycrate/obj/development/ -L/Users/imac27/Documents/Temp/mycrate/obj/development/ -L/users/imac27/.config/alire/cache/dependencies/gnat_native_12.2.1_77267eb1/lib/gcc/x86_64-apple-darwin19.6.0/12.2.0/adalib/ /users/imac27/.config/alire/cache/dependencies/gnat_native_12.2.1_77267eb1/lib/gcc/x86_64-apple-darwin19.6.0/12.2.0/adalib/libgnat.a -Wl,-rpath,@executable_path/..//obj/development -Wl,-rpath,@executable_path/../../../..//.config/alire/cache/dependencies/gnat_native_12.2.1_77267eb1/lib/gcc/x86_64-apple-darwin19.6.0/12.2.0/adalib -o /Users/imac27/Documents/Temp/mycrate/bin//mycrate
error: Command ["gprbuild", "-s", "-j0", "-p", "-P", "/Users/imac27/Documents/Temp/mycrate/mycrate.gpr"] exited with code 4
error: Compilation failed.

**Comment by mosteo** [comment]
Our expert on macOS is @simonjwright , he might have the answer to this.

Could you post the output of `alr version` to see if anything rings wrong?

**Comment by simonjwright** [comment, marked as answer]
The problem is that, with Sonoma, Apple have provided an updated Xcode version 15.0 (and the Command Line Tools (CLTs); @Darce42, please let us know which you’re using) which breaks the linking process.

You could confirm this by this command, which should work:
```
$ alr build -- -largs -Wl,-ld_classic
```

Unfortunately, you can’t reinstall the Command Line Tools version 14.3 on Sonoma, because it’s "too early". The same probably applies to Xcode.

What I _could_ do is provide you with the `ld` binary from Version 14.3, with instructions on how to install it;  but

* I’m not sure what Apple would have to say about that, though both you and I are licensed to use it
* I use the CLTs, so if you’re using Xcode I’d be taking an (informed!) guess as to where the replacement `ld` would need to go
* this might cause difficulties doing software development using Apple tools (clang, Swift...)

**Comment by Darce42** [comment]
Thank you for that suggestion. Simon has indeed replied with a comprehensive answer.

Darce

**Comment by Darce42** [comment]
Thank you very much Simon for that prompt and accurate answer.

My alire version is 1.2.2 with gnat_native=12.2.1 and gprbuild=22.0.1and Xcode is version 15.

I’m not keen to try any hacks which may work and which may break other tools - thank you for that warning.

Hopefully the formal fix will come in due course.

I tried your suggestion:

alr build -- -largs -Wl,-ld_classic

and I have been able to build and run a couple of programs successfully.

Best wishes

Darce
